src/qutip_qip/decompose/decompose_general_qubit_gate.py

```diff
@@ -0,0 +1,932 @@
+
+    # last set of gates with 0 as ctrl and 2 as target
+    V02_gate_list = _two_qubit_last_target(V, 0, 2)
+    for i in range(len(V02_gate_list)):
+        full_gate_list.append(V02_gate_list[i])
+    full_gate_list.append(Gate("X", targets=[2]))
+    return(full_gate_list)
+
+
+def _two_q_CNOT_gates(two_qubit_gate):
+    """Create full list of Pauli X and CNOT gates for final decomposition.
+    """
+    two_level_info = _decompose_to_two_level_arrays(
+        two_qubit_gate, 2, expand=False)
+
+    # create a list of non-trivial indices
+    non_trivial_indices = []
+    for i in range(len(two_level_info)):
+        non_trivial_indices.append(two_level_info[i][0])
+
+    # gate dict for two-level gates
+    gate_keys = list(range(len(two_level_info)))
+    full_gate_dict = {}
+
+    # gray code steps
+    full_gray_code_info = []
+    for i in range(len(two_level_info)):
+        ind1, ind2 = non_trivial_indices[i]
+        full_gray_code_info.append(_gray_code_gate_info(ind1, ind2, 2))
+
+    # add pauli gates
+    gray_code_ctrl_val_1 = {}
+    for i in range(len(full_gray_code_info)):
+        gray_pauli_i = _paulix_for_control_0(full_gray_code_info[i])
+        gray_code_ctrl_val_1[i] = gray_pauli_i
+
+    out = gray_code_ctrl_val_1
+
+    # create a dictionary of flattened lists
+    full_dict = {}
+    for i in range(6):
+        for_two_level_i = out[i]
+        if isinstance(for_two_level_i, dict):
+            full_dict[i] = [*out[i].values()]
+        elif isinstance(for_two_level_i, list):
+            if len(for_two_level_i) > 2:
+                full_dict[i] = for_two_level_i
+            elif len((for_two_level_i)) == 2:
+                forwa_seq = for_two_level_i[0]
+                back_seq = for_two_level_i[1]
+                forw_list = []
+                back_list = []
+                for_key_list = [*forwa_seq.keys()]
+                back_key_list = [*back_seq.keys()]
+                for j in for_key_list:
+                    if isinstance(forwa_seq[j], list):
+                        for k in range(len(forwa_seq[j])):
+                            forw_list.append(forwa_seq[j][k])
+                    else:
+                        forw_list.append(forwa_seq[j])
+                for p in back_key_list:
+                    if isinstance(back_seq[p], list):
+                        for q in range(len(back_seq[p])):
+                            back_list.append(back_seq[p][q])
+                    else:
+                        back_list.append(back_seq[p])
+                full_dict[i] = [forw_list, back_list]
+
+    # create a list of keys that will need CNOT gates or decomposition
+    cnot_ind = []
+    for i in range(6):
+        if len(full_dict[i]) == 2:
+            cnot_ind.append(i)
+    forw_ind_list = {}
+    backw_ind_list = {}
+    full_ind_list = []
+    for i in cnot_ind:
+        gray_code_info = full_dict[i]
+        forw = gray_code_info[0]
+        backw = gray_code_info[1]
```

Comment:
This will be changed after forward gray code is separated from the main two-level gate. Then backward gray code is just reversed order of forward gray code.